ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -547,6 +472,12 @@ private DataFlow::LocalSourceNode trackModuleAccess(Module m) {
   result = trackModuleAccess(m, TypeTracker::end())
 }
 
+pragma[nomagic]
+private predicate hasUserDefinedSelf(Module m) {
+  // cannot use `lookupSingletonMethod` due to negative recursion
+  singletonMethodOnModule(_, "new", m.getSuperClass*()) // not `getAnAncestor` because singleton methods cannot be included
```

Comment:
I suggest renaming this predicate to `hasUserDefinedNew`.

Could it be implemented in terms of `isUserDefinedNew` defined earlier in this file?

----

Also, we seem to lose some call edges due to `C.new` not being seen as an instance of `C` when `C` has an override of `self.new`, but where the implementation of `self.new` always returns `self.allocate`.

I think a reasonably simple fix would be to redefine this predicate to look for overriden `new` that _doesn't_ obviously return `self.allocate`:
```suggestion
  exists(DataFlow::MethodNode method |
    // not `getAnAncestor` because singleton methods cannot be included
    singletonMethodOnModule(method.asCallableAstNode(), "new", m.getSuperClass*()) and
    not method.getSelfParameter().getAMethodCall("allocate").flowsTo(method.getAReturningNode())
  )
```

But I'm OK with postponing this fix so as not to delay the PR any further.

Reply:
Good spot; thanks for checking. Let's fix follow-up.